lib/src/components/NxToast/NxToastProvider.tsx

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2019-present Sonatype, Inc.
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/.
+ */
+
+import React, { useCallback, useEffect, useState } from 'react';
+import ToastContext from './contexts';
+import NxToastContainer from './NxToastContainer';
+import { NxToastProviderProps, nxToastProviderPropTypes, ToastAddModel, ToastModel } from './types';
+
+let id = 1;
+
+const NxToastProvider = (props: NxToastProviderProps) => {
+  const { children } = props;
+  const [toasts, setToasts] = useState<ToastModel[]>([]);
+  const [activeElementBeforeToast, setActiveElementBeforeToast] = useState<HTMLElement | null>(null);
```

Comment:
Since the user can interact with the page behind the toast, I assume that when all toasts are closed we should return focus to the last non-toast element that was focused, which isn't necessarily the same element that was focused before any toasts were opened.  To accomplish this we can listen on `focusin` on the `document` and update this state value as we get focus events throughout the duration of toasts being open.  If that's done, the name of this variable should be adjusted slightly as well.

Reply:
> Since the user can interact with the page behind the toast, I assume that when all toasts are closed we should return focus to the last non-toast element that was focused, which isn't necessarily the same element that was focused before any toasts were opened. To accomplish this we can listen on `focusin` on the `document` and update this state value as we get focus events throughout the duration of toasts being open. If that's done, the name of this variable should be adjusted slightly as well.

Fixed here: [45fddad](https://github.com/sonatype/sonatype-react-shared-components/pull/688/commits/45fddad331825c8ac44e9b5eb350b20b3457f413)

I'm wondering if the callback function attached to `focusin` could slow down the page, since the [documentation](https://developer.mozilla.org/en-US/docs/Web/API/Element/focusin_event) mentions `focusin` bubbles out. The callback function is being called multiple times.  If the function could slow it down, maybe its best to use a memoized version of the callback, or to use the `focus` event instead.

Reply:
> I'm wondering if the callback function attached to focusin could slow down the page, since the [documentation](https://developer.mozilla.org/en-US/docs/Web/API/Element/focusin_event) mentions focusin bubbles out. The callback function is being called multiple times. If the function could slow it down, maybe its best to use a memoized version of the callback, or to use the focus event instead. Ross, if either option should be implemented, let me know.

The bubbling is crucial, without that we'd need to attach focus handlers to every element on the page :smile: .  The reason you are seeing the callback fire multiple times is because you need to register it in a `useEffect` (and unregister it in `useEffect`'s unmount callback) rather than directly in the `NxToastProvider` render function.  Memoization won't be relevant here since the registration will only be happening once per `NxToastProvider`.

Reply:
> The bubbling is crucial, without that we'd need to attach focus handlers to every element on the page 😄 . The reason you are seeing the callback fire multiple times is because you need to register it in a `useEffect` (and unregister it in `useEffect`'s unmount callback) rather than directly in the `NxToastProvider` render function. Memoization won't be relevant here since the registration will only be happening once per `NxToastProvider`.

I moved the the event into a useEffect here: [2b721d5](https://github.com/sonatype/sonatype-react-shared-components/pull/688/commits/2b721d5a60a656e7096b8c58312e9f513c95f870), which has greatly reduced the # of fires . I'm new to useEffect cleanups, so not sure if the cleanup is done correctly and cleanly

Reply:
> not sure if the cleanup is done correctly and cleanly

The cleanup looks right to me!  The only problem I see is that I don't think the dependency array is right. I what way does this rely on `activeElementNotToast`?

Reply:
> > not sure if the cleanup is done correctly and cleanly
> 
> The cleanup looks right to me! The only problem I see is that I don't think the dependency array is right. In what way does this rely on `activeElementNotToast`?

I was thinking that because adjustFocus is looking specifically for elements in focus that aren't toasts, and setting the activeElementNotToast state, it would make sense for the useEffect calling the function adjustFocus to be based on activeElementNotToast changing. However my initial instinct was to use [toasts], so I can change it to that if that's what you had in mind.

Reply:
> I was thinking that because adjustFocus is looking specifically for elements in focus that aren't toasts, and setting the activeElementNotToast state, it would make sense for the useEffect calling the function adjustFocus to be based on activeElementNotToast changing. However my initial instinct was to use [toasts], so I can change it to that if that's what you had in mind.

The dependency array should contain the props and other local variables that are used within the `useEffect` function and any functions that it calls/uses.  In this case, there aren't any.  Well there's `setActiveElementNotToast` but that's guaranteed to be stable across renders.  So there are no dependencies, and the list here should just be `[]`.

Reply:
> The dependency array should contain the props and other local variables that are used within the `useEffect` function and any functions that it calls/uses. In this case, there aren't any. Well there's `setActiveElementNotToast` but that's guaranteed to be stable across renders. So there are no dependencies, and the list here should just be `[]`.

That makes sense. It's updated here: [023ecc6](https://github.com/sonatype/sonatype-react-shared-components/pull/688/commits/023ecc699f98e21a167b5b93dbc55c12615be21e)